2022/day9/Program.fs

```diff
@@ -0,0 +1,91 @@
+﻿// https://adventofcode.com/2022/day/9
+open System.IO
+
+type Pos = struct
+    val X: int
+    val Y: int
+    new (x: int, y: int) = { X = x; Y = y }
+end
+
+// Create a new position by moving from the previous position.
+let movePos (x: int) (y: int) (p: Pos) = Pos(p.X + x, p.Y + y)
+
+// Parse a move sequence into a sequence of move function calls.
+// This will convert multi-cell moves into a sequence 1-cell of moves.
+let parseMoveSeq (move: string) =
+    match move.ToUpper().Split(' ') with
+    | [| "U"; count |] -> seq { for c in 1 .. (int count) -> movePos 0 -1 }
+    | [| "D"; count |] -> seq { for c in 1 .. (int count) -> movePos 0 +1 }
+    | [| "L"; count |] -> seq { for c in 1 .. (int count) -> movePos -1 0 }
+    | [| "R"; count |] -> seq { for c in 1 .. (int count) -> movePos +1 0 }
+    | _ -> []
+
+// Pull from the previous knot in the rope. If they are within 1 cell in
+// any direction, including diagonally, then no movement is needed.
+// Otherwise, move to follow the head, moving 0 or 1 position on each axis.
+let pullPos (h: Pos) (t: Pos) : Pos =
+    match (h.X - t.X, h.Y - t.Y) with
+    | (x,y) when abs x <= 1 && abs y <= 1 -> t
+    | (x,y) -> movePos (x.CompareTo(0)) (y.CompareTo(0)) t
+
+let processRopeMoves (knotCount: int) (verbosity: int) =
+    // Create the rope with all knot positions initialized to (0, 0).
+    let mutable knots = Array.create knotCount (Pos(0, 0))
```

Comment:
One thing that I would recommend trying to avoid is using mutable values.
When I was first learning F# it was hard for me to drop them, especially in a workload like this when it's:

- Set an initial value that will eventually hold end state
- Iterate over collection with making changes to that value
- Retrieve final value

The approaches that helped me thinking about this differently (hopefully they will be useful to you):
- convert it to recursion, at first your head will probably hurt (mine did), but it was a good mental exercise. I personally don't recommend using recursion too much other than for training tho
- think about how would I write it using Linq in C# trying to avoid mutation
- reading about collection methods in F# documentation, I especially recommend fold method https://fsharp.github.io/fsharp-core-docs/reference/fsharp-collections-listmodule.html#fold I've relied on it quite a lot when I was starting to learn

Of course F# also works nice with mutation and you can totally keep it like it is, but there are numerous benefits to doing it more pure way